Setting log level config from cloud.

noorhaq opened the pull request.

noorhaq committed:
Added Cloud to device configs.

noorhaq committed:
Fixed logs example typo.

noorhaq committed:
Updated partition.csv for esp32c6 coredump example.

noorhaq commented:
This has the cloud config settings.
A new mqtt.c file and function which uses events for the mqtt publish. More efficient and while idle it doesn't use CPU cycles. Because when idle the task will be in standby mode waiting for events. 
MQTT publish is created when MQTT is connected and is deleted when MQTT disconnects.

jmasek commented on esp_idf/spotflow/Kconfig.projbuild:
```diff
@@ -88,9 +82,27 @@ menu "Spotflow"
+    
+    config SPOTFLOW_DEFAULT_SENT_LOG_LEVEL
+        int "Spotflow default sent log level"
+        default 4
```
how is default log level done in esp? cannot use default value from esp?

noorhaq replied:
Default value is set to Info https://github.com/espressif/esp-idf/blob/fcae32885b0296b32044cb99ecbdc50d98dddb83/components/log/Kconfig.level#L3C5-L6C13

jmasek replied:
ok, why the default isn't `LOG_DEFAULT_LEVEL` 
``` 
 config SPOTFLOW_DEFAULT_SENT_LOG_LEVEL
        int "Spotflow default sent log level"
        default LOG_DEFAULT_LEVEL
```

noorhaq replied:
Yeah the default level is info so we can make that change I was testing it too so kept it at a higher debug level. But I can change this to info/default.

jmasek commented on esp_idf/spotflow/Kconfig.projbuild:
```diff
@@ -88,9 +82,27 @@ menu "Spotflow"
+    config SPOTFLOW_DEFAULT_SENT_LOG_LEVEL
+        int "Spotflow default sent log level"
+        default 4
+        range 0 6   
```
why 6 is maximum?

noorhaq replied:
Okay so it has 0 to 5 like 
0 None, 1 Error, 2 Warning, 3 Info, 4 Error, 5 Verbose, Added 6 in case we want to add critical or any other specific like Spotflow_log_debug here,

roberthusak replied:
I wouldn't do that - 6 would seem to be even more detailed than verbose which is IMO not needed.

noorhaq replied:
Changed.

jmasek commented on esp_idf/spotflow/examples/coredump/sdkconfig.defaults:
```diff
@@ -7,4 +7,6 @@ CONFIG_SPOTFLOW_DEVICE_ID=""
-CONFIG_PARTITION_TABLE_FILENAME="partitions.csv"
\ No newline at end of file
+CONFIG_PARTITION_TABLE_FILENAME="partitions.csv"
+CONFIG_LOG_MAXIMUM_EQUALS_DEFAULT=n
```
what this mean?

noorhaq replied:
Okay so it was a strange bug, if I don't set this value to n, by default it is y. Any configs I had added and declared in Kconfig were being ignored by the ESP_LOG subsystem.

jmasek replied:
if it is not working, shouldn't it be in kconfig - like spotflow `depends on !LOG_MAXIMUM_EQUALS_DEFAULT`

noorhaq replied:
Nope because it would still work for info, warning and error just not for the debug value like minimal compiled log level will decrease.
Other than that no change would be seen in the overall workflow.

noorhaq replied:
That's why I just disabled it because it was setting a few values and bypassing some which we require at a high level.

jmasek commented on esp_idf/spotflow/src/configs/spotflow_config.c:
```diff
@@ -0,0 +1,119 @@
+#include "logging/spotflow_log_backend.h"
+#include "logging/spotflow_log_cbor.h"
+#include "configs/spotflow_config.h"
+#include "configs/spotflow_config_cbor.h"
```
Is there a reason why we use `configs` in plural? other folders have singular - coredump, logging, net, buildId

noorhaq replied:
I planned to create a config/Spotflow_config.h so in that we can see which components were enabled and if any are missing like some other component is overwriting any of the configs we have set. 
We can generate a error at this point. So user can fix. 
And I couldn't think of a better name for the folder so I just made it configs. We can change that.

roberthusak replied:
I'd also prefer `config` as I understand it as just one configuration with possibly multiple configuration options.

Regarding checking the Kconfig values in `spotflow_config.h`, wouldn't it be cleaner (and possibly faster) to do it in CMake? We already check the MQTT buffer size in `CMakeLists.txt` so we can add other checks as necessary.

noorhaq replied:
It is but the common workflow is not to check in Cmake but in Kconfig so we can reuse. Like I tried to keep the configs same between our Zephyr and esp idf code.
So in future say we add support for Nuttx, or some other rtos or linux software. And their we might not be using cmake but some other. With the Kconfig.h we can port it easily just copy paste and change or remove the code we don't need. So minimal effort.

noorhaq replied:
Okay changed the folder name to config.

jmasek commented on esp_idf/spotflow/include/configs/spotflow_config_persistance.h:
```diff
@@ -0,0 +1,26 @@
+extern "C" {
+#endif
+
+#define SPOTFLOW_REPORTED_FLAG_MINIMAL_LOG_SEVERITY (1 << 0)
```
This is a mask used to signalize that variable `sent_log_level`, right? wouldn't be better to have name that contains the variable name? rather than minimal_log_severity?

noorhaq replied:
I kept the masks to the cbor key bindings like minimal log severity and kind of kept it in line with the zephyr one. Like I kept it in line with zephyr and the other variables set in the struct. To avoid confusion.

roberthusak replied:
In Zephyr, `struct spotflow_config_persisted_settings` contains the boolean field `contains_sent_log_level`, so I'd suggest to rename this constant to something like `SPOTFLOW_PERSISTED_SETTINGS_FLAG_SENT_LOG_LEVEL`. Or maybe something less wordy, but it confuses me that the current name references reported properties which are in different module and not directly related to persistence.

jmasek commented on esp_idf/spotflow/src/configs/spotflow_config.c:
```diff
@@ -0,0 +1,119 @@
+}
+
+/**
+ * @brief Initialize the sesson and send a intiail message with current configurations.
```
session

noorhaq replied:
I will fix this.

jmasek commented on esp_idf/spotflow/src/configs/spotflow_config_persistence.c:
```diff
@@ -0,0 +1,86 @@
+ *
+ * @param settings
+ */
+void spotflow_config_persistence_try_load(struct spotflow_config_persisted_settings* settings)
```
Shouldn't we write to flash only if the value has changed to limit the wear of flash memory?

jmasek commented on esp_idf/spotflow/src/configs/spotflow_config.c:
```diff
@@ -0,0 +1,119 @@
+
+	reported_msg->flags |= SPOTFLOW_REPORTED_FLAG_MINIMAL_LOG_SEVERITY;
+	reported_msg->minimal_log_severity =
+	    spotflow_cbor_convert_log_level_to_severity(sent_log_level);
```
Why are you converting it here and not in the function `spotflow_config_cbor_encode_reported`?

noorhaq replied:
It was just a preference to keep the encode clean. Like we save everything the way cloud likes before calling the encode which just encodes the data in struct we provide.

jmasek commented on esp_idf/spotflow/src/configs/spotflow_config_cbor.c:
```diff
@@ -0,0 +1,181 @@
+				return -1;
+			}
+			if (msg_type != UPDATE_DESIRED_CONFIGURATION_MESSAGE_TYPE) {
+				return -1;
```
couldyou add log for these errors, please?

noorhaq replied:
I think it would be pointless here as we are already doing a error check on the whole block side. 
We can add these logs to like SPOTFlOW_debug and such and add in like config.

jmasek commented on esp_idf/spotflow/src/configs/spotflow_config_persistence.c:
```diff
@@ -0,0 +1,86 @@
+#include "configs/spotflow_config_persistance.h"
+#include "logging/spotflow_log_backend.h"
+
+#define SPOTFLOW_STORAGE "Spotflow_S_L"
```
is there a reason for different casing between `SPOTFLOW_STORAGE` and `STORAGE_NAMESPACE`

noorhaq replied:
Fixed it, it was just a convention followed to keep the Namespace in like small case alphabets.

jmasek commented:
missing changelog

jmasek commented on esp_idf/spotflow/src/coredump/spotflow_coredump.c:
```diff
@@ -167,39 +168,45 @@ esp_err_t spotflow_coredump_backend(void)
-			vTaskDelay(pdMS_TO_TICKS(50)); //50 ticks wait
-		} while(rc < 0);
+			if (rc < 0) {
+				// vTaskDelay(pdMS_TO_TICKS(500));
```
remove if not used.

noorhaq committed:
Added MQTT 5 support. (#47)

jmasek commented on esp_idf/spotflow/include/net/spotflow_mqtt.h:
```diff
@@ -4,15 +4,23 @@
 #include "stdatomic.h"
 #include "mqtt_client.h"
 
+#define SPOTFLOW_MQTT_NOTIFY_COURDUMP (1 << 0) // Flag to trigger sending coredump
```
typo coredump

jmasek commented on esp_idf/spotflow/src/net/spotflow_mqtt.c:
```diff
@@ -4,178 +4,320 @@
-            }
+			// Try to send coredump messages first
+			if (notify_value & SPOTFLOW_MQTT_NOTIFY_COURDUMP) {
+				if(spotflow_coredump_send_message() == 5) {
```
magic number 5 - please use a predefined constant instead of a magic number

noorhaq replied:
Done.

jmasek commented on esp_idf/spotflow/src/net/spotflow_mqtt.c:
```diff
@@ -4,178 +4,320 @@
+		// Only proceed if the outbox_size i.e. current message is smaller than overall mqtt_buffer.
+		// 2️⃣ Wait until MQTT outbox has space (zero polling)
+		
+		while (esp_mqtt_client_get_outbox_size(spotflow_client) >=
```
how the default outbox size is defined? not sure if I understand the logic - if occupied space in the buffer is larger than `CONFIG_SPOTFLOW_CBOR_LOG_MAX_LEN`, stop pushing messages to client and wait for messages to be sent? is it correct?

noorhaq replied:
The mqtt buffer is used by not just our logging system but other services as well. By default it's value in esp idf a buffer size of 6144 bytes. Which you can adjust using the buffer size which we check in Cmakelist.txt if it's smaller than the CBOR_Max_len it forces the mqtt buffer to be equal to it. 
So yes if the buffer has some data which isn't sent and it's greater than the data we are about to add it skips and waits until it clears because we are not clearing the event here. Until the message is sent. 
And if smaller it adds the message to the outbox buffer i.e data being transmitted and moves on by clearing the event and the next things make sure whether the data was sent or not.

jmasek commented on esp_idf/spotflow/src/net/spotflow_mqtt.c:
```diff
@@ -4,178 +4,320 @@
+ * @param qos
+ * @return int
+ */
+int spotflow_mqtt_publish_messgae(const char* topic, const uint8_t* data, int len, int qos)
```
typo message

jmasek commented on esp_idf/spotflow/src/logging/spotflow_log_backend.c:
```diff
@@ -125,3 +125,28 @@ int spotflow_log_backend(const char* fmt, va_list args)
+		esp_log_level_set("*", level);  // Set log level for all
+    } else {
+        // Otherwise, set the log level for the specified tag
+        esp_log_level_set("*", level);  // Set log level for the provided tag
```
shouldn't be the first parameter tag?

noorhaq replied:
It should be but we are not receiving a tag currently so I have kept it like this. When we start receiving the tag we have to convert it to str and add it here. 
So kind of a placeholder. I will add a comment here.

roberthusak commented on esp_idf/spotflow/include/configs/spotflow_config_persistence.h:
Shouldn't we use "persist**e**nce" instead of "persist**a**nce" (for both `.c` and `.h` files)?

noorhaq replied:
I used zephyr code as base so the files, function names and arguments same as them. We can change it.

roberthusak replied:
Using the Zephyr code as a base is okay, the comment addresses just the typo (`a` instead of `e`) in these two file names (`spotflow_config_persistance.h` and `spotflow_config_persistance.c`).

noorhaq replied:
Yep i will fix this.

roberthusak commented on esp_idf/spotflow/src/configs/spotflow_config.c:
```diff
@@ -0,0 +1,119 @@
+	int rc = spotflow_config_prepare_pending_message(&reported_msg);
+	if (rc < 0) {
+		SPOTFLOW_LOG(
+		    "Failed to prepare initial reported configuration response message: %d", rc);
```
Please apply this minor fix - it's not a response. My bad, it's in Zephyr as well - can you fix it there as well?
```suggestion
		    "Failed to prepare initial reported configuration message: %d", rc);
```

roberthusak commented on esp_idf/spotflow/include/configs/spotflow_config.h:
```diff
@@ -0,0 +1,18 @@
+
+void spotflow_config_init(void);
+int spotflow_config_init_session(void);
+void spotflow_config_desired_message(const uint8_t* payload, int len);
```
I suggest adding a verb describing what the function does - e.g., `spotflow_config_handle_desired_message`.

noorhaq replied:
Again we can do that but we should keep the change in sync with the zephyr code base as well.

roberthusak replied:
This function is called `handle_desired_msg` in Zephyr code base. Just adding the prefix `spotflow_config_` because it's not static in ESP-IDF seems consistent to me. And `message` vs. `msg` is IMO just a detail, feel free to use either of them (Zephyr API often uses `msg` so that's why we used it there as well).

noorhaq replied:
I was just using a naming convention to know where the file is as at the start I wasn't planning to make it static. 
Like if I connect the handle_message to some other event based so that's why at the time i just named it and went past it without editing.
I think here it won't matter much if we use either. So let's go with the current one.

roberthusak replied:
It's OK that it isn't static. The original comment just addresses that the name of the function lacks a **verb** - "spotflow_config_desired_message" seems like a name of a structure whereas "spotflow_config_**handle**_desired_message" is a proper function name. So please just rename it.

roberthusak commented on esp_idf/spotflow/src/configs/spotflow_config.c:
```diff
@@ -0,0 +1,119 @@
+void spotflow_config_desired_message(const uint8_t* payload, int len)
+{
+	struct spotflow_config_desired_msg desired_msg;
+	SPOTFLOW_LOG("Deconding Payload\n");
```
```suggestion
	SPOTFLOW_LOG("Decoding payload\n");
```

roberthusak commented on esp_idf/spotflow/Kconfig.projbuild:
```diff
@@ -88,9 +82,27 @@ menu "Spotflow"
+            - 1 ERROR, send LOG_ERR
+            - 2 WARNING, send LOG_WRN in addition to previous level
+            - 3 INFO, send LOG_INF in addition to previous level
+            - 4 DEBUG, send LOG_DBG in addition to previous level
```
These levels should be updated for ESP-IDF.

roberthusak commented on esp_idf/spotflow/include/configs/spotflow_config_cbor.h:
```diff
@@ -0,0 +1,56 @@
+// Flags for reported
+#define SPOTFLOW_REPORTED_FLAG_MINIMAL_LOG_SEVERITY (1 << 0)
+#define SPOTFLOW_REPORTED_FLAG_COMPILED_MINIMAL_LOG_SEVERITY (1 << 1)
+#define SPOTFLOW_REPORTED_FLAG_ACKED_DESIRED_CONFIG_VERSION (1 << 2)
```
Why are the flags defined both in the enum and as macros?

roberthusak commented on esp_idf/spotflow/include/configs/spotflow_config_cbor.h:
```diff
@@ -0,0 +1,56 @@
+	uint64_t desired_config_version;
+};
+
+// clear a flag msg.flags &= ~SPOTFLOW_REPORTED_FLAG_ACKED_DESIRED_CONFIG_VERSION;
```
Is this comment necessary? I'd a expect a C programmer to know how to clear a flag.

noorhaq commented:
On the configs folder I remember it now when i changed it and rebuilt. During and after build in the build folder esp idf creates a folder called config which has the sdkconfig.h and other things so when I changed it back to config and checked it was causing linking issues because esp idf cmake treats directories with name config a little differently as per Chatgpt. I didn't Investigate it any further. 
@jmasek @roberthusak

noorhaq committed:
Fixed the case sensitive and spelling mistakes.

noorhaq committed:
Added queue empty number and remove unused code.

noorhaq committed:
Fixed typo in message.

noorhaq committed:
added spotflow_debug

roberthusak commented on esp_idf/spotflow/src/configs/spotflow_config_net.c:
```diff
@@ -0,0 +1,29 @@
+
+int spotflow_config_prepare_pending_message(struct spotflow_config_reported_msg* reported_msg)
+{
+	int rc = spotflow_config_cbor_encode_reported(reported_msg, pending_message_buffer,
```
Are we sure that there can't be a race condition on the shared buffer? For example, that `spotflow_config_prepare_pending_message` won't be executed while `spotflow_config_send_pending_message` is running?

noorhaq replied:
Yes. Because the mqtt task in esp idf is event based. When the event publish is running the incoming data is only saved in net recv buffers and isn't processed. And in case we ever receive it twice the data would be overwritten.

roberthusak commented on esp_idf/spotflow/include/configs/spotflow_config_options.h:
```diff
@@ -0,0 +1,20 @@
+void spotflow_config_init_sent_log_level_default();
+void spotflow_config_init_sent_log_level(uint8_t level);
+void spotflow_config_set_sent_log_level(uint8_t level);
+void spotflow_config_init_sent_log_level_default();
```
Duplicate?

jmasek replied:
not resolved @noorhaq

noorhaq replied:
Removed.

roberthusak commented on esp_idf/spotflow/src/configs/spotflow_config_persistence.c:
```diff
@@ -0,0 +1,86 @@
+{
+	esp_err_t err = nvs_flash_init();
+	if (err == ESP_ERR_NVS_NO_FREE_PAGES || err == ESP_ERR_NVS_NEW_VERSION_FOUND) {
+		ESP_ERROR_CHECK(nvs_flash_erase());
```
Isn't there a risk that we'll erase some user data unrelated to our module?

jmasek replied:
not resolved @noorhaq

noorhaq replied:
We are erasing only if the initialisation fails like if the nvs initialisation fails it means the data inside is already corrupted and we have a option in menuconfig to display a error to user and disable the NVS erase by default. 
They can enable it to erase if init fails but it is disabled by default. 
Because if the init fails meaning all the storage sectors or like the partition table for the nvs is corrupted. Due to some reason.
Here we are not checking our respective storage space like that comes afterwards.

roberthusak commented on esp_idf/spotflow/src/configs/spotflow_config_persistence.c:
```diff
@@ -0,0 +1,86 @@
+		return;
+	}
+
+	err = nvs_get_u8(spotflow_handle, SPOTFLOW_STORAGE, &settings->sent_log_level);
```
Shouldn't we rename `SPOTFLOW_STORAGE` to `SPOTFLOW_SENT_LOG_LEVEL_KEY` or something like that for better understandability?

jmasek replied:
not resolved @noorhaq

noorhaq replied:
Updated.

roberthusak commented on esp_idf/spotflow/include/coredump/spotflow_coredump.h:
```diff
@@ -11,7 +11,6 @@ extern "C" {
 bool spotflow_is_coredump_available(void);
 esp_err_t spotflow_coredump_backend(void);
 void spotflow_coredump_cleanup(void);
-
```
Typo? (the newline should be kept)

roberthusak commented on esp_idf/spotflow/Kconfig.projbuild:
```suggestion
        bool "To view the debug logs on the Spotflow side."
```

roberthusak commented on esp_idf/spotflow/src/logging/spotflow_log_backend.c:
```diff
@@ -125,3 +125,28 @@ int spotflow_log_backend(const char* fmt, va_list args)
+ * @param level 
+ * @param ... 
+ */
+void spotflow_log_backend_try_set_runtime_filter(uint8_t level, ...) {
```
Why are variadic arguments used there? Isn't it an overkill? AFAIK the tag isn't currently passed from any call site, so why is the function checking its presence at all? Seems like breaking the YAGNI principle.

noorhaq replied:
I just kept it here with a variadic argument. So we can keep the design consistent. When we add the tag level controls in the logs we can still use the code base without any changes. I can change it back.

roberthusak commented on esp_idf/spotflow/Kconfig.projbuild:
Please extend the label and help text so that the developer knows what's actually done when the option is enabled.

Also, if I understand it correctly, `esp_log_level_set` updates the log level not only for logs to the Spotflow module, but to all the log callbacks in the chain - such as when we're sending the logs to the cloud but they are also sent to UART. Therefore, the user might set the log level from the cloud to warning and more verbose messages would stop showing up on UART, which might not be expected. So, I'd prefer to have this option turned off by default. (It's not a problem in Zephyr, because the log level is adjusted there only for the single logging backend.)

noorhaq replied:
Okay yeah we can do that. On Kconfigs I was planning to go over in a seperate PR. Because that requires a lot of edits and back and forth on each option.

noorhaq replied:
It's in the Kconfig PR. Which will merge to this PR before merging to main.

roberthusak commented on esp_idf/spotflow/src/net/spotflow_mqtt.c:
```diff
@@ -4,178 +4,320 @@
+{
+	SPOTFLOW_LOG("MQTT Message Received on topic: %.*s", topic_len, topic);
+
+	// Compare topic exactly
```
"Exactly" sounds like we are checking the exact match, but we're checking just the prefix (because the actual topic will be longer due to the workspace and device IDs)., so I recommend updating the comment accordingly.

Also, the condition should check the prefix and not just containment - I suppose replacing `!= NULL` by `== topic` should solve that.

noorhaq replied:
Okay I added the strncmp instead of strstr which should do the following.

noorhaq committed:
Updated Kconfig.

noorhaq committed:
Addressed persistence changes.

noorhaq committed:
Changed initial config log message.

noorhaq committed:
Addressing some PR comments.

noorhaq committed:
Added config changes.

noorhaq committed:
Removed the extra arguments in the spotflow_log_backend_try_set_runtime_filter

noorhaq committed:
Added functionality to only write to NVS if the log level is changed.

noorhaq committed:
Updated Changelog.

roberthusak commented on esp_idf/spotflow/include/configs/spotflow_config_cbor.h:
```diff
@@ -0,0 +1,45 @@
+#define SPOTFLOW_DESIRED_FLAG_MINIMAL_LOG_SEVERITY (1 << 0)
+
+// Flags for reported
+#define SPOTFLOW_PERSISTED_SETTINGS_FLAG_SENT_LOG_LEVEL (1 << 0)
```
Please move this flag to `spotflow_config_persistence.h`, because persistence is unrelated to CBOR encoding. Furthermore, it's now unclear which flags can be used in the field `uint8_t flags` of `struct spotflow_config_persisted_settings`.

noorhaq replied:
Yep I will try to limit it somehow. I am using flags instead of bool to reduce struct padding like here the struct size will be smaller and we can add future expandability with just added flags. 
I will convert it to enums or some other. 
I placed these flags here in cbor instead of the respective files I think in case we add pre compilation Kconfigs to restrict some functionality so we won't have to deal with include errors as cbor will already have them and cbor is checking them. Again not sure i will give it another look.

roberthusak replied:
Using flags is OK 👍 I used boolean bitfields in the Zephyr module, but there seem to be some objective reasons to prefer flags instead. My concern isn't with using flags in general but that it should be apparent which flags are supposed be used in which structure. Using enums might help, but even properly named constants should suffice.

I probably see the problem now - why did you remove `SPOTFLOW_REPORTED_FLAG_MINIMAL_LOG_SEVERITY` from this file and replaced it with `SPOTFLOW_PERSISTED_SETTINGS_FLAG_SENT_LOG_LEVEL`? Although they have the same value, they are different things - the former should be used with `spotflow_config_reported_msg.flags` while the latter with `spotflow_config_persisted_settings.flags`.

By using the same value for both there's an artificial coupling between CBOR encoding and configuration persistence, which are two separate things. Therefore, as you wrote, we wouldn't be able to skip compilation of persistence because CBOR encoding code uses this constant.

So, please move the definition of `SPOTFLOW_PERSISTED_SETTINGS_FLAG_SENT_LOG_LEVEL` to `spotflow_config_persistence.h`, re-introduce `SPOTFLOW_REPORTED_FLAG_MINIMAL_LOG_SEVERITY` here, and use each flag with its respective `struct`.

I know it seems like a minor issue, but writing clean code now may save us many headaches in the future 😅

noorhaq replied:
Added using typedef. Do check it out.

noorhaq committed:
Fixed persistence file.

noorhaq committed:
Updated spotflow_config function name.

noorhaq committed:
Updated the config flags.

jmasek commented:
> On the configs folder I remember it now when i changed it and rebuilt. During and after build in the build folder esp idf creates a folder called config which has the sdkconfig.h and other things so when I changed it back to config and checked it was causing linking issues because esp idf cmake treats directories with name config a little differently as per Chatgpt. I didn't Investigate it any further. @jmasek @roberthusak

what about folder name `configuration`?

noorhaq committed:
Kconfig structure and menuoptions. (#51)

We can discuss and add the Kconfig structure in their respective files
and can discuss if we need to modify or change any in their respective
files. Or move some to the main file.

---------

Co-authored-by: Robert Husák <robert.husak@spotflow.io>

noorhaq committed:
Merge branch 'main' into feature/esp_idf_config

noorhaq committed:
Updated idf.py dependency file.

noorhaq committed:
Fixed config dublicate issues.
